Locale support for p:spinner

rrsln opened the pull request:
Locale support for p:spinner #453

rlytcrx committed:
Locale support for p:spinner

cnsgithub reviewed (commented):
@rrsln Are there really use cases where you need to specify `decimalSeparator` and `thousandSeparator` in a way that differs from the locale's default?

cnsgithub commented on src/main/resources/META-INF/primefaces-p.taglib.xml:
```diff
@@ -23902,6 +23902,22 @@
+        </attribute>
+		<attribute>
+            <description>
+                <![CDATA[Decimal separator char. Default is taken from the view's locale.]]>
```
`Character` or `String`? `DecimalFormatSymbols` uses `Character`, so this should be sufficient for us as well?

rrsln replied:
Character is enough for me. However I think that string is more flexible (for example someone may need to use space + separator) and it is consistent with p:inputNumber.

cnsgithub commented on docs/7_1/components/spinner.md:
```diff
@@ -35,6 +35,8 @@ min | null | Double | Minimum boundary value
 prefix | null | String | Prefix of the input
 suffix | null | String | Suffix of the input
+decimalSeparator | Default is taken from the view's locale | String | Decimal separator char.
+thousandSeparator | Default is taken from the view's locale | String | Thousand separator char.
```
I would prefer `groupingSeparator` instead of `thousandSeparator`, see `DecimalFormatSymbols`.

rrsln replied:
we use "thousandSeparator" in p:inputNumber. Using "thousandSeparator" there and "groupingSeparator" here is confusing.

cnsgithub replied:
Agreed, should be consistent. Didn't know that.

tandraschko commented:
i dont like the new style of:

`ComponentUtils.getValue(getStateHelper(), "thousandSeparator",
            () -> LocaleUtils.getThousandSeparator(getFacesContext()))`

we should leave it as it is and everywhere the same.

rrsln commented:
> 
> 
> @rrsln Are there really use cases where you need to specify `decimalSeparator` and `thousandSeparator` in a way that differs from the locale's default?

well, locale's default is sufficient for me. However it is easy to imagine usecase when there is a need to use something different (for example when you select locale from the combobox or would like to use "space + separator" rather than simply "separator".

cnsgithub commented:
> it is easy to imagine usecase

Sorry, cannot imagine. :alien:  I personally would not support those weird, non-standard formattings. Anyways, it should be consistent with the rest of PrimeFaces.

rrsln commented:
> 
> 
> i dont like the new style of:
> 
> `ComponentUtils.getValue(getStateHelper(), "thousandSeparator", () -> LocaleUtils.getThousandSeparator(getFacesContext()))`
> 
> we should leave it as it is and everywhere the same.

It is possible to use:
`new Supplier() {
   bla bla bla
}`
instead but in my opinion it is more readable with lambda and functional interfaces have been mostly introduced for lambdas. BTW we already use lambdas in PrimeApplicationContext

tandraschko commented:
It's not about the lambda, why not just use:

return (String) getStateHelper().eval(PropertyKeys.thousandSeparator, LocaleUtils.getThousandSeparator(getFacesContext()));

because LocaleUtils.getThousandSeparator is slow?

rrsln commented:
I doubt that LocaleUtils.getThousandSeparator is slow. For some reason it was written that way for InputNumber and I didn't want to change that. Also it doesn't harm when some method is not invoked assuming that readability is not sacrificed.

tandraschko commented:
Yep, i see.
Also created a feature request for JSF 3.x, so we can use a lambda for the future: https://github.com/eclipse-ee4j/mojarra/issues/4568

tandraschko commented:
Would be great if you can do following refactoring:
1) Move InputNumber get/setDecimalSeparator get/setThousandSeparator to InputNumberBase
2) Create PropertyKeys for this 2 properties instead of Strings and also use it in the getter/setter
3) Rename ComponentUtils#getValue - it's to generic for such a generic class. Maybe something evalLazy or something... no idea. There should be a reference to "stateHelper#eval" in the name

tandraschko commented:
it also conflicts now

rlytcrx committed:
Merge remote-tracking branch 'upstream/master' into PF453

# Conflicts:
#	docs/7_1/components/spinner.md

rrsln committed:
decimalSeparator/thousandSeparator code improvements

rrsln committed:
fix checkstyle warnings

rrsln commented:
fixed all comments. Renamed ComponentUtils#getValue to ComponentUtils#eval, it seems the best option to me.However feel free to rename if you believe that some other name is better.

tandraschko commented:
Looks good for me now
Can someone retest it?

melloware commented:
@tandraschko from my testing this PR looks good I didn't find anything wrong.
